src/producer/base_producer.rs

```diff
@@ -206,23 +213,80 @@ impl<'a, K: ToBytes + ?Sized, P: ToBytes + ?Sized> BaseRecord<'a, K, P, ()> {
     }
 }
 
+unsafe extern "C" fn partitioner_cb<Part: Partitioner, C: ProducerContext<Part>>(
+    topic: *const RDKafkaTopic,
+    keydata: *const c_void,
+    keylen: usize,
+    partition_cnt: i32,
+    rkt_opaque: *mut c_void,
+    _msg_opaque: *mut c_void,
+) -> i32 {
+    let topic_name = CStr::from_ptr(rdsys::rd_kafka_topic_name(topic));
+    let topic_name = str::from_utf8_unchecked(topic_name.to_bytes());
+
+    let is_partition_available = |p: i32| rdsys::rd_kafka_topic_partition_available(topic, p) == 1;
+
+    let key = if keydata.is_null() {
+        None
+    } else {
+        Some(slice::from_raw_parts(keydata as *const u8, keylen))
+    };
+
+    let producer_context = &mut *(rkt_opaque as *mut C);
+
+    match producer_context.get_custom_partitioner() {
+        None => panic!("custom partitioner is not set"),
```

Comment:
Hmm; is it possible to validate that the custom partitioner is set before the callback is called? Or is this a "will never happen" to satisfy the type system?

I'm not a fan of panicing in a library. What is the effect if the callback was set but no actual partitioner was configured? Can we set an error state somewhere that's bubbled out to the caller to handle? If it's basically a config error I think logging an error is better than panicing.

Reply:
Nothing stops author of producer context to e.g. return None from get_custom_partitioner() during 2nd and next call so in theory it can happen but it’s rather contrived case.

My first version there was to return PARTITION_UA (-1) but it leads to “message timeout” errors when message.timeout.ms elapses so for me that was more confusing than simply panicking right away.